### PR TITLE
feat: add human-readable duration format support for reservation timeouts

### LINKED ISSUE #121: reserveration timeout should be possible for Minutes, Hours, or Days
I may be wrong, but my understanding is that currently the reservation timeout has to be set in seconds.  This can be unwieldy when setting a reservation for several days or even when having do the math for several hour reservations.  It would be nice to use the same conventions as sleep uses:

30m (30 minutes)
5h (5 hours)
4d (4 days)

### PR DESCRIPTION
**Pull request by gajeshbhat**
## Description

This PR adds support for human-readable duration formats (e.g., `2h30m`, `5h`, `30m`) for reservation timeouts in job definitions and API requests, while maintaining full backward compatibility with the existing integer seconds format.

**Problem solved:** Users currently need to manually calculate seconds for reservation timeouts (e.g., calculating that 2.5 hours = 9000 seconds), making job definitions less readable and error-prone.

**Implementation approach:** 
- Added a duration parser in `testflinger-common` that handles various time units (s/m/h/d)
- Added `DurationField` to server schema for automatic parsing and validation
- **Server-only implementation** to avoid snapcraft packaging limitations with CLI cross-directory dependencies
- All components gracefully fall back to existing integer parsing for backward compatibility

## Resolved issues

Fixes #121 (CERTTF-586)

## Documentation
- Existing `docs/reference/test-phases.rst` correctly documents server-side duration format usage
- All public-facing changes are documented with clear examples showing job YAML usage
- Comprehensive tests included covering:
  - Duration parsing (valid/invalid formats)
  - Server schema validation
  - Backward compatibility scenarios

## Web service API changes

**Schema changes:**
- Modified `ReserveData.timeout` field to accept both integers and duration strings
- Added `DurationField` for automatic parsing: `"2h30m"` → `9000` seconds
- **Versioning:** No new endpoints - existing `/v1/job` endpoint enhanced
- **Backward compatibility:** All existing integer timeout values continue to work unchanged
- **Authorization:** No changes to existing authorization requirements
- **Performance:** No database query changes - only input validation enhancement

**Request/Response examples:**
```
// New duration format (converted to seconds internally)
{"reserve_data": {"timeout": "2h30m"}}  → timeout: 9000

// Existing formats still work
{"reserve_data": {"timeout": 3600}}     → timeout: 3600
{"reserve_data": {"timeout": "3600"}}   → timeout: 3600

// Invalid formats return 422 with a clear error message
{"reserve_data": {"timeout": "invalid"}} → 422 "Invalid duration format"
```

## Tests

**Testing performed:**
1. **Unit tests:** Duration parsing logic with comprehensive edge cases
2. **Schema validation:** Server accepts/rejects appropriate formats with clear error messages
3. **Backward compatibility:** All existing integer timeout formats continue to work
4. **Integration testing:** Full job submission workflow with duration formats

**Test commands to verify:**
```bash
# Test duration parsing
cd common && uv run pytest src/testflinger_common/tests/test_duration.py -v

# Test server schema validation
cd server && uv run pytest src/testflinger/tests/test_duration_schema.py -v

# Run all component tests
cd common && uvx --with tox-uv tox
cd server && uvx --with tox-uv tox
cd cli && uvx --with tox-uv tox
```

**Examples of supported formats:**
- `30s`, `30sec` (30 seconds)
- `30m`, `30min` (30 minutes)  
- `5h`, `5hour` (5 hours)
- `4d`, `4day` (4 days)
- `2h30m` (2 hours 30 minutes)
- `1d5h30m` (1 day 5 hours 30 minutes)

**Usage in job definitions:**
```yaml
job_queue: myqueue
provision_data:
  url: http://image.url
reserve_data:
  ssh_keys:
    - "lp:user"
  timeout: 2h30m 
```

## Notes
This implementation focuses on server-side parsing to avoid snapcraft packaging limitations while still providing the core benefit of human-readable duration formats in job definitions and API requests.

### CODE REVIEW

File: cli/pyproject.toml [hunk 1]
```diff
@@ -12,6 +12,7 @@ dependencies = [
   "pyjwt>=2.10.1",
   "pyyaml>=6.0.2",
   "requests>=2.32.3",
+  "testflinger-common",
```
[the remark this answers is not in the document]
Comment (pedro-avalos) [reply cli/pyproject.toml:15]: I think we should just have the server-side implementation for now

[the remark this answers is not in the document]
Comment (gajeshbhat) [reply cli/pyproject.toml]: Thank you. I have updated the changes. Kindly take a look and let me know if it works. We might have to create another issue off of #121 or update it to later support `--timeout` cli argument once the snapcraft packaging constraints are removed/reworked. Cheers :)

File: common/src/testflinger_common/duration.py [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+
+    # Minutes
+    if seconds >= 60:
+        minutes = seconds // 60
+        parts.append(f"{minutes}m")
+        seconds %= 60
+
+    # Seconds
+    if seconds > 0:
+        parts.append(f"{seconds}s")
```
Comment (pedro-avalos) [inline common/src/testflinger_common/duration.py:190]: What do you think about simplifying this with a loop:

```python
units = [
    (86400, "d"),
    (3600, "h"),
    (60, "m"),
    (1, "s"),
]
parts = []

for divisor, suffix in units:
    if seconds >= divisor:
        num = seconds // divisor
        parts.append(f"{num}{suffix}")
        seconds %= divisor
```

File: common/src/testflinger_common/duration.py [hunk 3]
```diff
@@ -0,0 +1,192 @@
...
+
+
+class DurationParseError(ValueError):
+    """Raised when a duration string cannot be parsed."""
+
+    pass
+
+
+def parse_duration(duration: Union[str, int]) -> int:
+    """Parse a duration string or integer into seconds.
```
Comment (pedro-avalos) [inline common/src/testflinger_common/duration.py:32]: Thank you for the detailed docstrings. I have a minor nitpick: We tend to use reStructuredText format for our docstrings, though. Could you update the docstrings to this format?

**Review by pedro-avalos — CHANGES_REQUESTED** [review]
Thank you for updating the PR. I just had a chance to look over your changes and have left some comments

**Comment by pedro-avalos** [comment]
Also, make sure to sign the CLA with your GitHub email: https://ubuntu.com/legal/contributors

File: common/src/testflinger_common/duration.py [hunk 4]
```diff
@@ -0,0 +1,192 @@
...
+        if duration < 0:
+            raise DurationParseError("Duration cannot be negative")
+        return duration
+
+    if not isinstance(duration, str):
+        raise DurationParseError(
+            f"Duration must be string or int, got {type(duration)}"
+        )
+
+    duration = duration.strip()
```
[the remark this answers is not in the document]
Comment (gajeshbhat) [reply common/src/testflinger_common/duration.py]: Good point. I've combined the `.strip()` and `.lower()` calls into a single operation: `duration.strip().lower()`.

File: common/src/testflinger_common/duration.py [hunk 5]
```diff
@@ -0,0 +1,192 @@
...
+    try:
+        value = int(duration)
+        if value < 0:
+            raise DurationParseError("Duration cannot be negative")
+        return value
+    except ValueError:
+        pass
+
+    # Check for negative numbers in duration strings
+    if duration.strip().startswith("-"):
```
[the remark this answers is not in the document]
Comment (gajeshbhat) [reply common/src/testflinger_common/duration.py]: Since `duration` was already stripped earlier in the function, the additional `.strip()` call is redundant. I've removed it to clean up the code.

File: common/src/testflinger_common/duration.py (hunk 2 above)
Comment (gajeshbhat) [reply common/src/testflinger_common/duration.py]: That's a neat approach. Thank you for recommending that. The loop approach is much cleaner and more maintainable than the repetitive if-statements.

File: common/src/testflinger_common/duration.py (hunk 3 above)
Comment (gajeshbhat) [reply common/src/testflinger_common/duration.py:32]: Thank you for pointing it out. Updated.

**Comment by gajeshbhat** [comment]
> Also, make sure to sign the CLA with your GitHub email: https://ubuntu.com/legal/contributors

I have signed the CLA by submitting the form using the provided link. Thank you, @pedro-avalos  for taking the time to review my PR. Please let me know if you have any further questions. Cheers.

File: common/src/testflinger_common/duration.py [hunk 6]
```diff
@@ -0,0 +1,182 @@
...
+        "hours": 3600,
+        "d": 86400,
+        "day": 86400,
+        "days": 86400,
+    }
+
+    for num_str, unit in matches:
+        num = int(num_str)
+        if num < 0:
+            raise DurationParseError("Duration components cannot be negative")
```
[the remark this answers is not in the document]
Comment (gajeshbhat) [reply common/src/testflinger_common/duration.py:135]: You're absolutely right! Removed the redundant negative check in the parsing loop since the regex pattern `(\d+)` only matches positive digits, making the negative check unreachable code.

File: common/src/testflinger_common/tests/test_duration.py [hunk 7]
```diff
@@ -0,0 +1,41 @@
+# Copyright (C) 2025 Canonical
```
[the remark this answers is not in the document]
Comment (gajeshbhat) [reply common/src/testflinger_common/tests/test_duration.py:1]: Moved tests to component level for consistency. Changed from `common/src/testflinger_common/tests/` to `common/tests/` to match the project structure used by other components.

File: server/src/testflinger/tests/test_duration_schema.py [hunk 8]
```diff
@@ -0,0 +1,49 @@
+# Copyright (C) 2025 Canonical
```
[the remark this answers is not in the document]
Comment (gajeshbhat) [reply server/src/testflinger/tests/test_duration_schema.py:1]: Instead of creating a separate test file, I added practical integration tests to `server/tests/test_v1.py` using parametrized tests as you suggested. The tests now validate actual job submission with different timeout formats (1800, "1800s", "30m") and verify invalid formats are rejected with 422 status. Do you think this is reasonable?

File: server/pyproject.toml [hunk 9]
```diff
@@ -89,4 +92,4 @@ ignore = "H023,H021"
 indent = 2
 
 [tool.uv.sources]
-testflinger-common = { path = "../common", editable = true }
\ No newline at end of file
+testflinger-common = { path = "../common" }
```
[the remark this answers is not in the document]
Comment (gajeshbhat) [reply server/pyproject.toml:95]: Updated the server's testflinger-common dependency to use `editable = true` for better development workflow as suggested.
